integration/tests/cook/test_basic.py

```diff
@@ -818,28 +818,57 @@ def test_group_kill_multi(self):
-        util.wait_until(query, util.all_instances_killed)
-        # Now try to kill the group again
-        # (ensure it still works when there are no live jobs)
-        util.kill_groups(self.cook_url, [group_uuid])
+        try:
+            jobs, resp = util.submit_jobs(self.cook_url, job_spec, 10, groups=[group_spec])
+            self.assertEqual(resp.status_code, 201)
+            # Wait for some job to start
+            def query():
+                return util.group_detail_query(self.cook_url, group_uuid)
+            util.wait_until(query, util.group_some_job_started)
+            # Now try to kill the whole group
+            util.kill_groups(self.cook_url, [group_uuid])
+            # Wait for all the jobs to die
+            # Ensure that each job Failed (because we killed it)
+            def query():
+                return util.query_jobs(self.cook_url, True, job=jobs)
+            util.wait_until(query, util.all_instances_killed)
+        finally:
+            # Now try to kill the group again
+            # (ensure it still works when there are no live jobs)
+            util.kill_groups(self.cook_url, [group_uuid])
+
+    def test_group_change_retries(self):
+        group_spec = self.minimal_group()
+        group_uuid = group_spec['uuid']
+        job_spec = {'group': group_uuid, 'command': f'sleep 1'}
+        try:
+            jobs, resp = util.submit_jobs(self.cook_url, job_spec, 10, groups=[group_spec])
+            self.assertEqual(resp.status_code, 201)
+            # wait for some job to start
+            def group_query():
+                return util.group_detail_query(self.cook_url, group_uuid)
+            util.wait_until(group_query, util.group_some_job_started)
+            # kill the whole group
+            util.kill_groups(self.cook_url, [group_uuid])
+            # wait for all the jobs to die
+            def jobs_query():
+                return util.query_jobs(self.cook_url, True, job=jobs)
+            util.wait_until(jobs_query, util.all_instances_done)
+            # retry all jobs in the group
+            resp = util.retry_jobs(self.cook_url, retries=2, groups=[group_uuid])
+            # wait for some job to start
+            util.wait_until(group_query, util.group_some_job_started)
+            # ensure none of the jobs are still in a failed state
+            job_data = util.query_jobs(self.cook_url, job=jobs)
+            self.assertEqual(200, job_data.status_code)
+            for job in job_data.json():
+                if job['status'] not in ['waiting', 'running']:
+                    job_details = f'Job details: {json.dumps(job, sort_keys=True)}'
+                    self.assertEqual('completed', job['status'], job_details)
+                    self.assertEqual('success', job['state'], job_details)
+        finally:
+            # ensure that we don't leave a bunch of jobs running/waiting
+            util.kill_groups(self.cook_url, [group_uuid])
```

Comment:
I broke this test into two different cases: `test_group_change_killed_retries` and `test_group_change_completed_retries`. The latter tests a mixed case, where some are running, some are completed, and some are still waiting. I'll add a couple more variants when I implement the "failed-only" flag for retries.

Reply:
Does `test_group_change_completed_retries` actually ensure that some jobs are running and some are still waiting when it does the group retry? It seems like it only ensures that at least one job has completed (but it could be that all are completed, or that the non-completed ones are all waiting or all running). I'm not necessarily advocating for making the test more complex, but maybe we should make its intent more clear with a comment. Also, the name is a little confusing.

Reply:
I'll change the name to `test_group_change_retries`, and add a comment to clarify what the expected state is after the `wait_until` call with the some-job-completed predicate.